server/fleet/app.go

```diff
@@ -139,6 +150,14 @@ type EnrichedAppConfig struct {
 	Logging         *Logging               `json:"logging,omitempty"`
 }
 
+// UnmarshalJSON implements the json.Unmarshaler interface.
+// It is explicitly defined to prevent promoted methods from embedded structs
+// from taking over JSON-serialization.
+func (e *EnrichedAppConfig) UnmarshalJSON(b []byte) error {
+	type alias *EnrichedAppConfig
+	return json.Unmarshal(b, alias(e))
```

Comment:
It won't! here's a small repro to illustrate https://go.dev/play/p/vNwPiA9y4KG

I learned this recently, and my understanding is that methods are not assigned to the new type, from the spec:

> A defined type may have [methods](https://go.dev/ref/spec#Method_declarations) associated with it. It does not inherit any methods bound to the given type, but the [method set](https://go.dev/ref/spec#Method_sets) of an interface type or of elements of a composite type remains unchanged
> https://go.dev/ref/spec#Type_definitions

Reply:
Yeah, sorry it's hard to properly explain, that's not quite the case I'm worried about - I know that a new type doesn't inherit the methods defined _on the type on which it is based_, but in this case, the method we care about is defined on the embedded `AppConfig` struct, which I believe would still be "inherited" by the new `alias` type?

In other words, in this spec paragraph:

> It does not inherit any methods bound to the given type, but the [method set](https://go.dev/ref/spec#Method_sets) of an interface type or of elements of a composite type remains unchanged

I'm worried about "the method set of elements of a composite type" that remain unchanged. I'll try to update the playground example to show what I mean.

Reply:
Something like that should've set the "Y" field in T2 (which is T1 without custom unmarshalling), but it ended up calling Embedded.UnmarshalJSON instead and did not unmarshal anything in any field (barring any mistakes in my quick code :D ): https://go.dev/play/p/5orTQrdSzFq

Reply:
@mna thank you very much for building that playground.

Our situation is _slightly_ bit different, check `T3` [here](https://go.dev/play/p/MkRQjP0lb5k)

The reason this "works" (our tests pass): both `EnrichedAppConfig` and `appConfigResponse` are used to present app configs, where we don't need at all backwards compatibility (and thus we don't _really_ need `UnmarshalJSON`). This is super confusing and wasn't clear to me until we had this discussion.

I can think of three solutions:

1. Leave it like that with a more accurate comment
2. Do a smarter `Unmarshal`ling, pseudo code:

```go
var appConfig AppConfig
json.Unmarshal(b, &appConfig)

var extras struct {
	UpdateInterval  *UpdateIntervalConfig  `json:"update_interval,omitempty"`
	Vulnerabilities *VulnerabilitiesConfig `json:"vulnerabilities,omitempty"`
	License         *LicenseInfo           `json:"license,omitempty"`
	Logging         *Logging               `json:"logging,omitempty"`
}
json.Unmarshal(b, &extras)

e.AppConfig = appConfig
e.UpdateInterval = extras.UpdateInterval
// ...
```
3. Figure out an smarter way to add backwards-compat, maybe doing it at the JSON-deserialization level is not a good idea to start with.

Reply:
PS: this is perhaps a clearer way to accomplish the same effect for `1`? https://go.dev/play/p/ILOo_MEGijW

Reply:
Oh ok, thanks for your patience there! I mean, if it works it's fine to me, I was just worried that it might not do exactly what we want it to. Yeah maybe a comment saying not to worry :D I can definitely see myself in 6 months seeing this and raising the same concern!

Reply:
no, thank you! you made me realize this wasn't working as I assumed. Will add a comment, but I'm also totally fine revisiting the approach.

Reply:
as usual, you were right. This ended up causing problems down the line. I used a variant of option `2` described above and implemented a custom unmarshaler for those types. Will mark this as solved.